### PR TITLE
feat: add support for `ServiceWorkers`

### LINKED ISSUE #11: [Feature] support `ServiceWorker`
Yes, I know, they are different things. **But**...
- ... [serviceworker-loader](https://github.com/markdalgleish/serviceworker-loader), is dead
- ... and it is basically a copy of an old version of this, with minor adjustments
- ... because most of the webpack code is the same for a service worker (emit a new file, allow instanciation), only the instanciation helper is different
- ... and they don't share the "worker" part of their name for no reason :)

### PR DESCRIPTION
**Pull request by AsaAyers**
Fixes #11

### CODE REVIEW

**Comment by opatut** [comment]
May I, at this point, say thank you for implementing this? Cheers :tada: :+1:

**Comment by markdalgleish** [comment]
I'd love for this to get merged in so I can deprecate https://github.com/markdalgleish/serviceworker-loader.

**Comment by AsaAyers** [comment]
This PR is pretty straight forward, but I think the more important part is #15. I don't know what useful things people would do with a service worker that don't include the list of assets from webpack that need to be cached. #15 introduces a magic `__assets__` array.

#15 functions, but really needs some help before it's ready to merge. Any help would be greatly appreciated.

**Comment by AsaAyers** [comment]
@markdalgleish Are you able to help get both of my PRs merged? #15 still needs some help. I've had these PRs open for 3 months now and you're the only webpack member I've seen a response from.

**Comment by patrickdet** [comment]
this loader works well for me so far. love it. thx @AsaAyers

File: index.js [hunk 1]
```diff
@@ -42,12 +42,17 @@ module.exports.pitch = function(request) {
 		if(err) return callback(err);
 		if (entries[0]) {
 			var workerFile = entries[0].files[0];
-			var constructor = "new Worker(__webpack_public_path__ + " + JSON.stringify(workerFile) + ")";
-			if(query.inline) {
-				constructor = "require(" + JSON.stringify("!!" + path.join(__dirname, "createInlineWorker.js")) + ")(" +
+			var constructor
+			if(query.service) {
+				constructor = "navigator.serviceWorker.register(__webpack_public_path__ + " + JSON.stringify(workerFile) + ", options);"
```
Comment (halkeye) [inline index.js]: Can you add detection here? Worker is pretty much defined everywhere, serviceWorker isn't

maybe something like
`navigator.serviceWorker ? navigator.serviceWorker.register(__webpack_public_path__ + " + JSON.stringify(workerFile) + ", options) : Promise.reject('No support')`

Comment (AsaAyers) [inline index.js]: ok, now it rejects if `navigator.serviceWorker` is missing.

**Comment by AsaAyers** [comment]
@markdalgleish I see you're a webpack member now. Are you able to help with this and #15?

**Comment by HyeonuPark** [comment]
Is there some example to use this with SW? I don't think the way below is correct.
```js
myServiceWorker = require('worker!./sw.js')

navigator.serviceWorker.register(myServiceWorker)
```

**Comment by AsaAyers** [comment]
If you want a regular worker the code is: 

```js
myServiceWorker = require('worker!./sw.js')

new myServiceWorker()
```

On my branch you create a service worker using:

```js
myServiceWorker = require('worker?service!./sw.js')

new myServiceWorker()
```

File: index.js [hunk 2]
```diff
@@ -42,12 +42,19 @@ module.exports.pitch = function(request) {
 		if(err) return callback(err);
 		if (entries[0]) {
 			var workerFile = entries[0].files[0];
-			var constructor = "new Worker(__webpack_public_path__ + " + JSON.stringify(workerFile) + ")";
-			if(query.inline) {
-				constructor = "require(" + JSON.stringify("!!" + path.join(__dirname, "createInlineWorker.js")) + ")(" +
+			var constructor
+			if(query.service) {
+				constructor = "('serviceWorker' in navigator)" +
+				"? navigator.serviceWorker.register(__webpack_public_path__ + " + JSON.stringify(workerFile) + ", options);" +
```
Comment (suprememoocow) [inline index.js]: I'm getting a syntax error in the generated code on this line, caused by the semi-colon before the ternary operator colon:

If you remove it, it seems to work correctly. 

```
? navigator.serviceWorker.register(__webpack_public_path__ + " + JSON.stringify(workerFile) + ", options) " 
```

**Comment by michael-ciniawsky** [comment]
Duplicate of #32 ?

**Comment by AsaAyers** [comment]
I'm making https://github.com/pulls more useful to me today by closing my old PRs.

**Pull request closed without merging**